File: bindings/nim/src/proven/safe_string.nim
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
#
# Safe string operations for security-sensitive contexts.
# Thin wrapper over libproven FFI -- all logic lives in Idris.

import std/options
import lib_proven

proc isValidUtf8*(value: string): bool =
  ## Check if a string contains valid UTF-8 bytes.
  if value.len == 0:
    return true
  let res = provenStringIsValidUtf8(unsafeAddr value[0], csize_t(value.len))
  if res.status == PROVEN_OK:
    return res.value
  false

proc escapeSql*(value: string): Option[string] =
  ## Escape single quotes for SQL strings.
  ## Note: Use parameterized queries when possible.
  if value.len == 0:
    return some("")
  let res = provenStringEscapeSql(unsafeAddr value[0], csize_t(value.len))
  if res.status == PROVEN_OK and res.value != nil:
    let escaped = $res.value
    provenFreeString(res.value)
    return some(escaped)
  none(string)

proc escapeHtml*(value: string): Option[string] =
  ## Escape HTML special characters to prevent XSS attacks.
  if value.len == 0:
    return some("")
  let res = provenStringEscapeHtml(unsafeAddr value[0], csize_t(value.len))
  if res.status == PROVEN_OK and res.value != nil:
    let escaped = $res.value
    provenFreeString(res.value)
    return some(escaped)
  none(string)

proc escapeJs*(value: string): Option[string] =
  ## Escape JavaScript special characters.
  if value.len == 0:
    return some("")
  let res = provenStringEscapeJs(unsafeAddr value[0], csize_t(value.len))
  if res.status == PROVEN_OK and res.value != nil:
    let escaped = $res.value
    provenFreeString(res.value)
    return some(escaped)
  none(string)

File: bindings/php/src/Proven/SafeString.php
<?php
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// SafeString - FFI wrapper for proven_string_* functions.
// All computation delegated to libproven; no logic reimplemented here.

declare(strict_types=1);

namespace Proven;

/**
 * Safe string operations for escaping and sanitization.
 *
 * Every method calls the corresponding libproven FFI function.
 * Returns null on error.
 */
final class SafeString
{
    /**
     * Check if a byte string is valid UTF-8.
     */
    public static function isValidUtf8(string $data): ?bool
    {
        $result = FFI::getLib()->proven_string_is_valid_utf8($data, strlen($data));
        if ($result->status !== 0) {
            return null;
        }
        return (bool) $result->value;
    }

    /**
     * Escape string for SQL (single quotes).
     */
    public static function escapeSql(string $value): ?string
    {
        return self::extractString(
            FFI::getLib()->proven_string_escape_sql($value, strlen($value))
        );
    }

    /**
     * Escape string for HTML (< > & " ').
     */
    public static function escapeHtml(string $value): ?string
    {
        return self::extractString(
            FFI::getLib()->proven_string_escape_html($value, strlen($value))
        );
    }

    /**
     * Escape string for JavaScript (within quotes).
     */
    public static function escapeJs(string $value): ?string
    {
        return self::extractString(
            FFI::getLib()->proven_string_escape_js($value, strlen($value))
        );
    }

    /**
     * Extract string from StringResult, free C memory, return null on error.
     */
    private static function extractString(mixed $result): ?string
    {
        if ($result->status !== 0 || $result->value === null) {
            return null;
        }
        $str = \FFI::string($result->value, $result->length);
        FFI::getLib()->proven_free_string($result->value);
        return $str;
    }
}
